File: src/WebUI/ClientApp/src/services/PortfolioService.ts
import axiosInstance from "../axiosInstance";
import { Portfolio, PortfolioModel } from "../models/models";

export class PortfolioService {
  CreatePortfolio = async (portfolio: Portfolio) => {
    const res: PortfolioModel = await (
      await axiosInstance.post("api/Portfolio/CreatePortfolio", portfolio)
    ).data;
    return res;
  };

  FindOrCreate = async (containerId: number) => {
    const res: PortfolioModel = await (
      await axiosInstance.put(`api/Portfolio/FindOrCreate/${containerId}`)
    ).data;

    return res;
  };

  DeletePortfolio = async (portfolio: Portfolio) => {
    const res: PortfolioModel = await (
      await axiosInstance.delete(`api/Portfolio/DeletePortfolio${portfolio.id}`)
    ).data;
    return res;
  };
}

File: src/Infrastructure/Services/EntityServices/PortfolioService.cs
using AutoMapper;
using HomiePages.Application.RepositoryInterfaces;
using HomiePages.Application.ServiceInterfaces;
using HomiePages.Domain.Entities;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using System.Threading.Tasks;

namespace HomiePages.Infrastructure.Services.EntityServices
{
    public class PortfolioService : ServiceBaseHelper<Portfolio>, IPortfolioService
    {
        private readonly IRepositoryWrapper _repo;
        private readonly IMapper _mapper;

        public PortfolioService(IRepositoryWrapper repo, IMapper mapper, IHttpContextAccessor httpContextAccessor) 
            : base(repo, repo.Portfolios, httpContextAccessor)
        {
            _repo = repo;
            _mapper = mapper;
        }

        public Portfolio GetPortfolio(Expression<Func<Portfolio, bool>> expression)
        {
            return GetBy(expression).Include(t => t.Equities).FirstOrDefault();
        }

        public Portfolio FindOrCreate(string userId, long containerId)
        {
            var portfolio = GetPortfolio(p => p.ContainerId == containerId);

            if (portfolio != null)
            {
                return portfolio;
            }

            var newPortfolio = new Portfolio()
            {
                UserId = userId,
                ContainerId = containerId
            };

            _repo.Portfolios.Create(newPortfolio);
            var res = _repo.SaveChanges();

            return newPortfolio;
        }

        public bool CreatePortfolio(Portfolio portfolio)
        {
            _repo.Portfolios.Create(portfolio);
            return _repo.SaveChanges();
        }

        public bool DeletePortfolio(int portfolioId)
        {
            return DeleteById(portfolioId);
        }

    }
}
